### PR TITLE
Provisioning k8s cluster on amazon/aliyun with script

### LINKED ISSUE #57: Provisioning k8s cluster on amazon/aliyun
Provisioning k8s cluster on amazon and aliyun, with configurations like `accessId` and `accessSecrets`.

### PR DESCRIPTION
**Pull request by acezen**
## What do these changes do?
This pr provide a script to provisions k8s cluster on amazon/aliyun and output a kube config file.

Useage:
```bash
./scripts/launch_cluster.py --type aws --id your_access_key_id --secret your_access_key_secrete --region your_region_name --output kube_config_path
```
then the script will ask user to enter the information (eg. key_id, key_secret, region...) to use/create a k8s cluster.
## Related issue number

Fixes #57 

todo:
- [x] add some comments and revise the prompt input and output.

### CODE REVIEW

**Comment by siyuan0322** [comment]
Should we also adding a link to the README and GetStarted documentation on how to use this script? 

A humble opinion: Give a summary comment in the script on what this launcher is doing and what permissions are required. Because if the user need to pass in AccessKeyID and AccessKeySecret (which are very sensitive information) to your script, they may have concerns about what exactly you're doing with the script, e.g. worrying about evil things.

The purpose of this script and usage should be demonstrated in help strings of `-h` and `--help`.

File: README.md [hunk 1]
```diff
@@ -40,6 +40,23 @@ On Windows, you may want to install [Ubuntu](https://ubuntu.com/blog/ubuntu-on-w
 ./scripts/prepare_env.sh
 ```
 
+### Deploy with AWS/Aliyun
+
+In addition to local cluster setup script, we also provide a interactive script to set up a Kubernetes cluster on AWS or Aliyun. The script would output a kube config file of the Kubernetes cluster.
```
Comment (siyuan0322) [inline README.md:45]: output到哪里了？直接打出来还是 output to /path/to/.kubeconfig ？

File: scripts/launch_cluster.py [hunk 2]
```diff
@@ -0,0 +1,642 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Create a Kubernetes cluster on AWS/Aliyun using python sdk.  You should know that this
+# takes an absolute age, allow 20-30 mins.  That's just the speed stuff
```
Comment (siyuan0322) [inline scripts/launch_cluster.py:5]: `allow 20-30 mins` what does this mean.

File: scripts/launch_cluster.py [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+                    {
+                        "ParameterKey": "Subnets",
+                        "ParameterValue": ",".join(vpc_subnet_ids)
+                    }
+                ],
+                TimeoutInMinutes=15,
+                OnFailure='DELETE'
+            )
+
+            if response == None:
```
Comment (siyuan0322) [inline scripts/launch_cluster.py:319]: use `is` or `is not` to compare to `None`. `if response is None:`

File: scripts/launch_cluster.py [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+        click.echo("Write config map...")
+        worker_auth_file =  os.path.dirname(self._output_path) + "/aws-auth-cm.yaml"
+        with open(worker_auth_file, "w") as f:
+            f.write(config)
+        
+        # add the worker resource to cluster
+        resp = subprocess.call(["kubectl", "--kubeconfig=%s" % self._output_path, "apply",
+                                "-f", worker_auth_file])
+        if resp != 0:
+            click.echo("The kubectl command didn't work.")
```
Comment (siyuan0322) [inline scripts/launch_cluster.py:363]: consider use `[subprocess.run](https://docs.python.org/3/library/subprocess.html#subprocess.run)`? And save the output and errors by `stdout=subprocess.PIPE, stderr=subprocess.PIPE`, so if exception occurs you can get the error messages in exception. As mentioned in doc:
> If check is true, and the process exits with a non-zero exit code, a CalledProcessError exception will be raised. Attributes of that exception hold the arguments, the exit code, and stdout and stderr if they were captured.

**Review by lidongze0629 — COMMENTED** [review]
Do we support command line parameters?

```python3
./launch_cluster.py --cloud_type aws --access_id_key xxxx --region beijing
```
No, in addition to id, key and region, user still need to input other information like cluster name,  ecs instance type etc, it would be better use interactive input to get the information.

File: scripts/launch_cluster.py (hunk 2 above)
Comment (acezen) [reply scripts/launch_cluster.py:5]: it means the whole creation process would take 20-30 mins

**Comment by acezen** [comment]
> Do we support command line parameters?
> 
> ```python
> ./launch_cluster.py --cloud_type aws --access_id_key xxxx --region beijing
> ```
> 
> No, in addition to id, key and region, user still need to input other information like cluster name, ecs instance type etc, it would be better use interactive input to get the information.

the latest commit support **./launch_cluster.py --cloud_type aws --access_id_key xxxx --region beijing**

**Review by yecol — APPROVED** [review]
LGTM for the script. For the docs, maybe we can find a more suitable place for them.

File: README.md [hunk 5]
```diff
@@ -40,6 +40,23 @@ On Windows, you may want to install [Ubuntu](https://ubuntu.com/blog/ubuntu-on-w
...
 ```
 
+### Deploy with AWS/Aliyun
+
+In addition to local cluster setup script, we also provide a interactive script to set up a Kubernetes cluster on AWS or Aliyun. The script would output a kube config file of the Kubernetes cluster.
+You can use the script as follows or use `./script/launch_cluster.py --help` to get the useage.
+
+- AWS
+```bash
+pip3 install click yaml boto3
```
Comment (siyuan0322) [inline README.md:50]: pip package of yaml's name is actually **pyyaml**.

File: scripts/launch_cluster.py [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+            click.echo("Creating VPC stack...")
+            # Create VPC stack.
+            response = self._cf.create_stack(
+                StackName=vpc_name,
+                TemplateURL=self.vpc_template,
+                Parameters=[],
+                TimeoutInMinutes=15,
+                OnFailure='DELETE'
+            )
+            if response == None:
```
Comment (siyuan0322) [inline scripts/launch_cluster.py:209]: use `is None`.

File: scripts/launch_cluster.py [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+        click.echo("Write config map...")
+        worker_auth_file =  os.path.dirname(self._output_path) + "/aws-auth-cm.yaml"
+        with open(worker_auth_file, "w") as f:
+            f.write(config)
+        
+        # add the worker resource to cluster
+        try:
+            subprocess.run(["kubectl", "--kubeconfig=%s" % self._output_path, "apply",
+                           "-f", worker_auth_file], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        except Exception as e:
```
Comment (siyuan0322) [inline scripts/launch_cluster.py:361]: The error info is in `e.stderr`. ref:
https://github.com/alibaba/GraphScope/blob/10f36b1dee462bf3efe9439a354e4f9e6d9f6da1/k8s/precompile.py#L46-L47

File: scripts/launch_cluster.py [hunk 8]
```diff
@@ -0,0 +1,640 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Create a Kubernetes cluster on AWS/Aliyun using python sdk.  
+#
+# Notice: this script would require your AWS/Aliyun account's access_key_id and secret_access_key
+# to get access your clusters' information or create a Kubernetes cluster automatically, finally output
+# a kube config of cluster. 
+# 
+# This script assumes that kubectl and aws-iam-authenticator are in your PATH.
```
Comment (siyuan0322) [inline scripts/launch_cluster.py:10]: This should be checked before launcher starts. If not exists, remind user to install them.

File: scripts/launch_cluster.py [hunk 9]
```diff
@@ -0,0 +1,640 @@
...
+                sys.exit(1)
+
+
+@click.command()
+@click.option("--type", type=click.Choice(["aws", "aliyun"], case_sensitive=False), 
+              help="Cloud type to launch cluster.")
+@click.option("--id", help="The access_key_id of cloud.")
+@click.option("--secret", help="The access_key_secret of cloud.")
+@click.option("--region", help="The region id.")
+@click.option("--output", help="The kube config file output path.", default=os.environ["HOME"] + "/.kube/config",
```
Comment (siyuan0322) [inline scripts/launch_cluster.py:617]: If user's home already contains a config, would you overwrite that?
If so, please make a backup and tell user where it is.

File: scripts/launch_cluster.py [hunk 10]
```diff
@@ -0,0 +1,640 @@
...
+            # Is it active? If so break out of loop
+            if status == "running":
+                break
+            elif status == "failed":
+                click.echo("Cluster create failed.")
+                sys.exit(1)
+            # Maybe give up after so many goes.
+            cnt = cnt - 1
+            if cnt <= 0:
+                click.echo("Given up waiting for cluster to go RINNING.")
```
Comment (siyuan0322) [inline scripts/launch_cluster.py:607]: RINNING -> RUNNING?

**Review by siyuan0322 — APPROVED** [review]
Launch a cluster passed given all preparation is ok.
But look through all error handling in the code, if any error is encounter it just print a `failed` then exit, I'm afraid it's hard for user to know what's the root cause of the failure. This could be left for future enhancement of course.

In addition, please lint and format the code using flake8 and black. You can look the lint procedure of CI for reference.
https://github.com/alibaba/GraphScope/blob/10f36b1dee462bf3efe9439a354e4f9e6d9f6da1/.github/workflows/ci.yml#L85-L90

File: scripts/launch_cluster.py [hunk 11]
```diff
@@ -0,0 +1,633 @@
...
+                sys.exit(1)
+
+
+@click.command()
+@click.option("--type", type=click.Choice(["aws", "aliyun"], case_sensitive=False), 
+              help="Cloud type to launch cluster.")
+@click.option("--id", help="The access_key_id of cloud.")
```
[the remark this answers is not in the document]
Comment (acezen) [reply scripts/launch_cluster.py]: add required=True to option that required.

File: scripts/launch_cluster.py [hunk 12]
```diff
@@ -0,0 +1,633 @@
...
+@click.option("--secret", help="The access_key_secret of cloud.")
+@click.option("--region", help="The region id.")
+@click.option("--output", help="The kube config file output path.", default=os.environ["HOME"] + "/.kube/config",
+              show_default=True)
+def launch(type, id, secret, region, output):
+    """Utility script to launch cluster on AWS or Aliyun and output kube config file."""
+
+    if type == "aws":
+        launcher = AWSLauncher(id, secret, region, output)
```
[the remark this answers is not in the document]
Comment (acezen) [reply scripts/launch_cluster.py:650]: done.

File: scripts/launch_cluster.py (hunk 6 above)
Comment (acezen) [reply scripts/launch_cluster.py]: fixed.

File: scripts/launch_cluster.py (hunk 7 above)
Comment (acezen) [reply scripts/launch_cluster.py]: fixed.

File: scripts/launch_cluster.py (hunk 9 above)
Comment (acezen) [reply scripts/launch_cluster.py:633]: done.

File: scripts/launch_cluster.py (hunk 10 above)
Comment (acezen) [reply scripts/launch_cluster.py]: fixed.

File: scripts/launch_cluster.py (hunk 8 above)
Comment (acezen) [reply scripts/launch_cluster.py:10]: done.

**Comment by acezen** [comment]
> Launch a cluster passed given all preparation is ok.
> But look through all error handling in the code, if any error is encounter it just print a `failed` then exit, I'm afraid it's hard for user to know what's the root cause of the failure. This could be left for future enhancement of course.
> 
> In addition, please lint and format the code using flake8 and black. You can look the lint procedure of CI for reference.
> https://github.com/alibaba/GraphScope/blob/10f36b1dee462bf3efe9439a354e4f9e6d9f6da1/.github/workflows/ci.yml#L85-L90

thanks for testing the script, already fix the comments.

**Pull request merged**